hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/HoodieSparkSqlWriter.scala

```diff
@@ -429,6 +416,40 @@ object HoodieSparkSqlWriter {
     }
   }
 
+  def deduceOperation(hoodieConfig: HoodieConfig, paramsWithoutDefaults : Map[String, String]): WriteOperationType = {
+    var operation = WriteOperationType.fromValue(hoodieConfig.getString(OPERATION))
+    // TODO clean up
+    // It does not make sense to allow upsert() operation if INSERT_DROP_DUPS is true
+    // Auto-correct the operation to "insert" if OPERATION is set to "upsert" wrongly
+    // or not set (in which case it will be set as "upsert" by parametersWithWriteDefaults()) .
+    if (hoodieConfig.getBoolean(INSERT_DROP_DUPS) &&
+      operation == WriteOperationType.UPSERT) {
+
+      log.warn(s"$UPSERT_OPERATION_OPT_VAL is not applicable " +
+        s"when $INSERT_DROP_DUPS is set to be true, " +
+        s"overriding the $OPERATION to be $INSERT_OPERATION_OPT_VAL")
+
+      operation = WriteOperationType.INSERT
+      operation
+    } else {
+      // if no record key, no preCombine, we should treat it as append only workload
+      // and make bulk_insert as operation type.
+      if (!paramsWithoutDefaults.containsKey(DataSourceWriteOptions.RECORDKEY_FIELD.key())
+        && !paramsWithoutDefaults.containsKey(DataSourceWriteOptions.PRECOMBINE_FIELD.key())
+        && !paramsWithoutDefaults.containsKey(OPERATION.key())) {
+        log.warn(s"Choosing BULK_INSERT as the operation type since auto record key generation is applicable")
+        operation = WriteOperationType.BULK_INSERT
+      }
+      // if no record key is set, will switch the default operation to INSERT (auto record key gen)
+      else if (!hoodieConfig.contains(DataSourceWriteOptions.RECORDKEY_FIELD.key())
+        && !paramsWithoutDefaults.containsKey(OPERATION.key())) {
+        log.warn(s"Choosing INSERT as the operation type since auto record key generation is applicable")
+        operation = WriteOperationType.INSERT
```

Comment:
so to help me confirm: when auto key gen is applicable (no record key and no operation), we enforce INSERT if precombine field is set, or BULK_INSERT if precombine field not set.

Reply:
i meant enforced was conditioned on "when auto key gen is applicable". My question was more on the choice of "INSERT" vs "BULK_INSERT": both have the same semantics, but here it chooses INSERT when users set precombine field and the other when not set, which I don't see the rationale behind. Shouldn't the choice btw those 2 based on some file sizing flag, rather than having precombine field or not?

Reply:
I see your point now. so, are you suggesting, we should check of explicit file sizing is overriden and non 0, we should go w/ insert, if not, we should go w/ bulk_insert ?

Reply:
yup it's about justifying the choice of INSERT based on other configs. making it depend on precombine causes confusion. file sizing makes more sense. A different point: can we have a unified helper to determine if auto keygen is applicable? and can this deduceOperation helper be moved to a common module so it's not limited to spark?

Reply:
hey @xushiyan : I do hear your point. Here we are mainly simplifying a user who is comes from writing to parquet table to hudi table. thats why we are looking into the mandatory fields like record key, precombine. Can we go ahead and land this in for 0.14.0. 
we can continue our discussion on how to simplify further. bcoz, very likely if someone is setting file sizing, they know hudi to some extent and not just trying to replace df.write.parquet w/ df.write.hudi. we can jot down few more use-cases and come up w/ a holistic plan

Reply:
ok understood. can we just deduce operation as bulk insert then? keep precombine field out of the deducing logic, as it causes confusion